llvm/lib/Target/AIE/AIECombinerHelper.cpp

```diff
@@ -1109,3 +1111,210 @@ void llvm::applyPadVector(MachineInstr &MI, MachineRegisterInfo &MRI,
+///                   %209(<16 x s64>), %68(s32)
+///  %219:_(<8 x s64>) = G_INTRINSIC intrinsic(@llvm.aie2.ext.512.1024.acc),
+///                    %209(<16 x s64>), %93(s32)
+
+/// To convert to:
+///   %216:_(<8 x s64>) = COPY %206(<8 x s64>)
+///   %219:_(<8 x s64>) = COPY %208(<8 x s64>)
+bool llvm::matchExtractConcat(MachineInstr &MI, MachineRegisterInfo &MRI,
+                              const AIEBaseInstrInfo &TII,
+                              Register &MatchInfo) {
+
+  GIntrinsic &GMI = cast<GIntrinsic>(MI);
+  std::optional<const AIEBaseInstrInfo::VExtractOpInfo> ExtractOp =
+      TII.getVExtractOpInfo(GMI.getIntrinsicID());
+
+  if (!ExtractOp)
+    return false;
+
+  auto Cst = getIConstantVRegValWithLookThrough(
+      GMI.getOperand(ExtractOp->SubVectorIndex).getReg(), MRI);
+
+  if (!Cst)
+    return false;
+
+  const unsigned ExtractSize =
+      MRI.getType(MI.getOperand(0).getReg()).getSizeInBits();
+
+  MachineInstr &SrcMI =
+      *MRI.getVRegDef(GMI.getOperand(ExtractOp->Src).getReg());
+
+  Register SrcReg;
+  unsigned ConcatSize = 0;
+  const unsigned Index = Cst->Value.getZExtValue();
+  if (SrcMI.getOpcode() == TargetOpcode::G_CONCAT_VECTORS) {
+    SrcReg = SrcMI.getOperand(Index + 1).getReg();
+    ConcatSize = MRI.getType(SrcReg).getSizeInBits();
+
+  } else if (SrcMI.getOpcode() == TargetOpcode::G_INTRINSIC) {
+    GIntrinsic &SrcMIIntrinsic = cast<GIntrinsic>(SrcMI);
+
+    if (const auto ConcatOp =
+            TII.getVConcatOpInfo(SrcMIIntrinsic.getIntrinsicID())) {
+      SrcReg =
+          SrcMIIntrinsic.getOperand(Index + ConcatOp->FirstOperand).getReg();
+      ConcatSize = MRI.getType(SrcReg).getSizeInBits();
+    }
+  }
+
+  MatchInfo = SrcReg;
+  return ConcatSize == ExtractSize;
+}
+
+void llvm::applyExtractConcat(MachineInstr &MI, MachineRegisterInfo &MRI,
+                              MachineIRBuilder &B, Register &MatchInfo) {
+  B.setInstrAndDebugLoc(MI);
+  Register DstReg = MI.getOperand(0).getReg();
+  Register SrcReg = MatchInfo;
+
+  B.buildCopy(DstReg, SrcReg);
+  MI.eraseFromParent();
+}
+
+/// Match something like this:
+/// %209:_(<16 x s32>) = G_INTRINSIC intrinsic(@llvm.aie2.concat.I512.I256),
+///         %95(<8 x s32>), %98(<8 x s32>)
+/// %252:_(<8 x s32>), %253:_(<8 x s32>) = G_UNMERGE_VALUES %209(<16 x s32>)
+///
+/// To convert to:
+/// 252:_(<8 x s32>) = COPY %95(<8 x s32>)
+/// 253:_(<8 x s32>) = COPY %98(<8 x s32>)
+bool llvm::matchUnmergeConcat(MachineInstr &MI, MachineRegisterInfo &MRI,
+                              const AIEBaseInstrInfo &TII,
+                              std::pair<MachineInstr *, unsigned> &MatchInfo) {
+
+  MachineInstr &SrcMI =
+      *MRI.getVRegDef(MI.getOperand(MI.getNumOperands() - 1).getReg());
+
+  if (SrcMI.getOpcode() != TargetOpcode::G_INTRINSIC)
+    return false;
+
+  GIntrinsic &GMI = cast<GIntrinsic>(SrcMI);
+
+  std::optional<const AIEBaseInstrInfo::VConcatOpInfo> ConcatOp =
+      TII.getVConcatOpInfo(GMI.getIntrinsicID());
```

Comment:
Curious: Why do you look at `G_CONCAT_VECTOR` as well in the combiner above, but not here?

Reply:
Good question! I was looking to patterns in several benchmarks and probably I was not able to see this pattern happening in any case.